### PR TITLE
BLOOM enhance alibi creation

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Thanks to @justheuristic 's contribution alibi tensor is better created/communicated during the forward pass. The tests seem to pass but still stays as an experimental feature.

cc @justheuristic

This probably will break with accelerate offloading because when initialising alibi tensor we do it only once at the beginning of the forward pass with the device of the first hidden states. In the previous version we used to dynamically change alibi's `device` which was fine for accelerate offloading

### CODE REVIEW

File: src/transformers/models/bloom/modeling_bloom.py [hunk 1]
```diff
@@ -359,12 +336,12 @@ def forward(
         output_attentions=False,
     ):
         # hidden_states: [batch_size, seq_length, hidden_size]
-        # repeat alibi tensor with the batch size
-        alibi = alibi.repeat(hidden_states.shape[0], 1, 1).to(hidden_states.device)
-
         # apply preprocessing if the input is padded
         if attention_mask is not None and 0 in attention_mask:
```
Comment (justheuristic) [inline src/transformers/models/bloom/modeling_bloom.py:340]:
```suggestion
        if attention_mask is not None:
```
I'd recommend not to check on 0 in attention mask: this triggers host-device synchronization, which slows down GPU and paralyzes TPU.

Comment (younesbelkada) [reply src/transformers/models/bloom/modeling_bloom.py]: Very good to know ! Thanks for the suggestion

File: src/transformers/models/bloom/modeling_bloom.py [hunk 2]
```diff
@@ -359,12 +336,12 @@ def forward(
         output_attentions=False,
     ):
         # hidden_states: [batch_size, seq_length, hidden_size]
-        # repeat alibi tensor with the batch size
-        alibi = alibi.repeat(hidden_states.shape[0], 1, 1).to(hidden_states.device)
-
         # apply preprocessing if the input is padded
-        if attention_mask is not None and 0 in attention_mask:
```
Comment (justheuristic) [inline src/transformers/models/bloom/modeling_bloom.py:366]: ... and here as well :)

Comment (younesbelkada) [reply src/transformers/models/bloom/modeling_bloom.py:366]: This code snippet has been already erased ! It was in the previous version

File: src/transformers/models/bloom/modeling_bloom.py [hunk 3]
```diff
@@ -359,12 +336,12 @@ def forward(
...
-            alibi = pre_process_alibi_for_pad(alibi, attention_mask, self.num_heads)
+        if attention_mask is not None:
+            alibi = pre_process_alibi_for_pad(alibi, attention_mask)
+        # otherwise repeat alibi tensor with the batch size
+        else:
```
Comment (sgugger) [inline src/transformers/models/bloom/modeling_bloom.py:343]:
```suggestion
        else:
            # otherwise repeat alibi tensor with the batch size
```
Small nit

**Review by sgugger — APPROVED** [review]
Could you elaborate on what's breaking? I see the alibi tensor is created at each forward pass and is now put on the same device as the `hidden_states` which actually seems clean!

It doesn't break model parallelism/offload as when you call each submodule with the alibi tensor as inputs, each submodule will place it on the right device before executing the forward. I checked the corresponding tests and they all pass on this branch.

**Comment by younesbelkada** [comment]
Great thanks ! I think that you are right :) will merge it as soon as the lights are all green 🟢

**Comment by younesbelkada** [comment]
It looks like a bad rebase happened, moved the PR at: #17866
